local/sdk-sriov/pkg/deviceplugin/server.go

```diff
@@ -0,0 +1,167 @@
+	ctx           context.Context
+}
+
+// StartServer creates a new SR-IOV forwarder device plugin server and starts it
+func StartServer(ctx context.Context, config *ServerConfig, manager Manager) error {
+	logFmt := "StartServer: %v"
+
+	s := &devicePluginServer{
+		resourceName:  resourceNamePrefix + config.ResourceName,
+		resourceCount: config.ResourceCount,
+		hostBaseDir:   config.HostBaseDir,
+		hostPathEnv:   config.HostPathEnv,
+		ctx:           ctx,
+	}
+
+	logrus.Infof(logFmt, "starting server")
+	socket, err := manager.StartDeviceServer(s.ctx, s)
+	if err != nil {
+		logrus.Errorf(logFmt, "error starting server")
+		return err
+	}
+
+	logrus.Infof(logFmt, "registering server")
+	if err := s.register(manager, socket); err != nil {
+		logrus.Errorf(logFmt, "error registering server")
+		return err
+	}
+
+	if resetCh, err := manager.MonitorKubeletRestart(s.ctx); err == nil {
+		go func() {
+			logrus.Infof(logFmt, "start monitoring kubelet restart")
+			defer logrus.Infof(logFmt, "stop monitoring kubelet restart")
+			for {
+				select {
+				case <-s.ctx.Done():
+					return
+				case _, ok := <-resetCh:
+					if !ok {
+						return
+					}
+					logrus.Infof(logFmt, "re registering server")
+					if err := s.register(manager, socket); err != nil {
+						logrus.Errorf(logFmt, fmt.Sprint("error re registering server: ", err))
+						return
+					}
+				}
+			}
+		}()
+	} else {
+		logrus.Warnf(logFmt, "error monitoring kubelet restart")
+	}
+
+	return nil
+}
+
+func (s *devicePluginServer) register(manager Manager, socket string) error {
+	return manager.RegisterDeviceServer(s.ctx, &pluginapi.RegisterRequest{
+		Version:      pluginapi.Version,
+		Endpoint:     socket,
+		ResourceName: s.resourceName,
+	})
+}
+
+func (s *devicePluginServer) GetDevicePluginOptions(_ context.Context, _ *pluginapi.Empty) (*pluginapi.DevicePluginOptions, error) {
+	return &pluginapi.DevicePluginOptions{}, nil
+}
+
+func (s *devicePluginServer) ListAndWatch(_ *pluginapi.Empty, server pluginapi.DevicePlugin_ListAndWatchServer) error {
+	logFmt := "devicePluginServer(ListAndWatch): %v"
+
+	resp := &pluginapi.ListAndWatchResponse{
+		Devices: make([]*pluginapi.Device, s.resourceCount),
+	}
+	for i := 0; i < s.resourceCount; i++ {
+		resp.Devices[i] = &pluginapi.Device{
+			ID:     fmt.Sprintf("%s/%v", s.resourceName, i),
+			Health: pluginapi.Healthy,
+		}
+	}
+
+	for {
+		if err := server.Send(resp); err != nil {
+			logrus.Errorf(logFmt, "server unavailable")
+			return err
+		}
+		select {
+		case <-s.ctx.Done():
+			logrus.Infof(logFmt, "server stopped")
+			return nil
+		case <-time.After(kubeletNotifyDelay):
+		}
+	}
+}
+
+func (s *devicePluginServer) Allocate(_ context.Context, request *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	response := &pluginapi.AllocateResponse{
+		ContainerResponses: make([]*pluginapi.ContainerAllocateResponse, len(request.ContainerRequests)),
+	}
+
+	for i := range request.ContainerRequests {
```

Comment:
Please increment count +1

Reply:
Now there is a `DeviceMonitor` and so we update count every times it fires.